cmd/gateway-manta.go

```diff
@@ -0,0 +1,368 @@
+func (t *tritonObjects) ListObjects(bucket, prefix, marker, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
+	ctx := context.Background()
+	objs, err := t.client.Dir().List(ctx, &storage.ListDirectoryInput{
+		DirectoryName: path.Join(mantaDefaultRootStore, bucket, prefix),
+	})
+	if err != nil {
+		return result, err
+	}
+
+	var blObjects []ObjectInfo
+	var blPrefixes []string
+
+	for _, obj := range objs.Entries {
+
+		if obj.Type == "directory" {
+			blPrefixes = append(blPrefixes, fmt.Sprintf("%s/", obj.Name))
+		} else {
+			var bucketInfo ObjectInfo
+			bucketInfo.Name = obj.Name
+
+			pathway := path.Join(bucket, prefix)
+			objInfo, err := t.GetObjectInfo(pathway, obj.Name)
+			if err != nil {
+				return result, err
+			}
+
+			bucketInfo.ContentType = objInfo.ContentType
+			bucketInfo.ETag = objInfo.ETag
+			bucketInfo.ContentEncoding = objInfo.ContentEncoding
+			bucketInfo.Size = objInfo.Size
+			bucketInfo.ModTime = objInfo.ModTime
+
+			blObjects = append(blObjects, bucketInfo)
+		}
+	}
+
+	return ListObjectsInfo{
+		Prefixes: blPrefixes,
+		Objects:  blObjects,
+	}, nil
+}
+
+// GetObject - Reads an object from Manta. Supports additional parameters like
+// offset and length which are synonymous with HTTP Range requests.
+//
+// startOffset indicates the starting read location of the object.  length
+// indicates the total length of the object.
+//
+// https://apidocs.joyent.com/manta/api.html#GetObject
+func (t *tritonObjects) GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer) error {
+
+	ctx := context.Background()
+	output, err := t.client.Objects().Get(ctx, &storage.GetObjectInput{
+		ObjectPath: path.Join(mantaDefaultRootStore, bucket, object),
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(writer, output.ObjectReader)
+	defer output.ObjectReader.Close()
+	return nil
+}
+
+// GetObjectInfo - reads blob metadata properties and replies back ObjectInfo,
+// uses Triton equivalent GetBlobProperties.
+//
+// https://apidocs.joyent.com/manta/api.html#GetObject
+func (t *tritonObjects) GetObjectInfo(bucket, object string) (objInfo ObjectInfo, err error) {
+	var info ObjectInfo
+	ctx := context.Background()
+	obj, err := t.client.Objects().Get(ctx, &storage.GetObjectInput{
+		ObjectPath: path.Join(mantaDefaultRootStore, bucket, object),
+	})
+	if err != nil {
+		return info, err
+	}
+
+	info = ObjectInfo{
+		Bucket:      bucket,
+		ContentType: obj.ContentType,
+		Size:        int64(obj.ContentLength),
+		ModTime:     obj.LastModified,
+	}
+
+	return info, nil
+}
+
+// PutObject - Create a new blob with the incoming data, uses Triton equivalent
+// CreateBlockBlobFromReader.
+//
+// https://apidocs.joyent.com/manta/api.html#PutObject
+func (t *tritonObjects) PutObject(bucket, object string, data *HashReader, metadata map[string]string) (objInfo ObjectInfo, err error) {
+	ctx := context.Background()
+	var oInfo ObjectInfo
+
+	b, err := ioutil.ReadAll(data)
```

Comment:
This will run out of memory @stack72 for simple uploads like 2GB.  You can simply stream this properly directly.